File: 2024/02.py
from utils import aoc

def is_safe(report: list[int]) -> bool:
    differences = [b - a for a, b in zip(report, report[1:])]
    return (all(d > 0 for d in differences) or 
            all(d < 0 for d in differences)) and\
            all(1 <= abs(d) <= 3 for d in differences)

def is_safe_with_dampener(report: list[int]) -> bool:
    if is_safe(report): 
        return True
    return any(is_safe(report[:i] + report[i+1:]) for i in range(len(report)))

if __name__ == "__main__":
    input = aoc.get_input("02")
    reports = [list(map(int, report.split())) for report in input]
    print(sum(map(is_safe, reports)))
    print(sum(map(is_safe_with_dampener, reports)))

File: 2024/02.exs
defmodule DayTwo do
  defp parse_line(line) do
    line
    |> String.trim()
    |> String.split()
    |> Enum.map(&String.to_integer/1)
  end

  defp read_input do
    "./inputs/02.txt"
    |> File.stream!()
    |> Stream.map(&parse_line/1)
    |> Enum.to_list()
  end

  defp is_monotonic?(list, comparator) do
    Enum.chunk_every(list, 2, 1, :discard)
    |> Enum.all?(fn [a, b] -> comparator.(a, b) end)
  end

  defp valid_deltas?(list) do
    Enum.chunk_every(list, 2, 1, :discard)
    |> Enum.all?(fn [a, b] -> abs(b - a) in 1..3 end)
  end

  defp is_safe?(list) do
    (is_monotonic?(list, &Kernel.</2) or is_monotonic?(list, &Kernel.>/2)) and valid_deltas?(list)
  end

  defp is_safe_with_dampener?(list) do
    Enum.with_index(list)
    |> Enum.any?(fn {_, index} ->
      is_safe?(List.delete_at(list, index))
    end)
  end

  def part_one() do
    read_input()
    |> Enum.filter(&is_safe?/1)
    |> Enum.count()
  end

  def part_two() do
    read_input()
    |> Enum.filter(&is_safe_with_dampener?/1)
    |> Enum.count()
  end
end

IO.inspect(DayTwo.part_one())
IO.inspect(DayTwo.part_two())
